File: ncarg2d/src/libncarg/support/icfell.f
C
C $Id: icfell.f,v 1.7 2008-07-27 00:17:29 haley Exp $
C                                                                      
C                Copyright (C)  2000
C        University Corporation for Atmospheric Research
C                All Rights Reserved
C
C The use of this Software is governed by a License Agreement.
C
      FUNCTION ICFELL (MESSG,NERRF)
C
        CHARACTER*(*) MESSG
C
C ICFELL (which stands for "I Check For Errors on Lower Level") is used
C to check for the occurrence of a recoverable error in a lower-level
C routine and (perhaps) to update the current error flag and message.
C
C The old value of the current error flag is returned as the value of
C the function ICFELL.  If that value is zero, nothing else has been
C done.  If the value of ICFELL is non-zero, the following actions have
C been taken:
C
C   If MESSG is blank, the current error message has not been changed.
C
C   If MESSG is non-blank and its length is 6 or less, it should be the
C   name of the routine referencing ICFELL; the current error message
C   has been altered by prepending first a slash and then the value of
C   MESSG.  (Note that, if an error occurs several levels deep, the
C   effect of using ICFELL at each level is, effectively, to generate
C   traceback information in the error message.)
C
C   If MESSG is non-blank and its length is 7 or greater, its value
C   has become the new value of the current error message and the
C   previous error message has been printed.  This is used at the
C   beginning of an NCAR Graphics routine to check for an outstanding
C   error that the user has not recovered from and to ensure that the
C   message for the outstanding error gets printed.
C
C   If the expression NERRF has the value zero, the current error
C   flag has not been changed.
C
C   If the expression NERRF has a non-zero value, the value of the
C   current error flag has been made equal to that value.
C
C An example:  Assume that the routine "A" calls the routine "B" and
C that "B" detects an error and calls SETER with error number "32" and
C error message "B - ERROR HAS OCCURRED".  If recovery mode is not in
C effect, SETER prints the error message and STOPs, but, if recovery
C mode is in effect, control returns from "SETER" to "B" and thence to
C "A".  At that point, the statement
C
C   IF (ICFELL('A',13).NE.0) RETURN
C
C detects the fact that an error has occurred in "B" and results in a
C return from "A" to whatever routine called it.  It also changes the
C current error message to read "A/B - ERROR HAS OCCURRED" and changes
C the error number from "32" to "13".
C
C Another example:  Assume that the NCAR Graphics routine "A" is called
C when recovery mode is set and that it detects an error, calls SETER,
C and RETURNs to the user.  If the user neglects to check the error
C state and calls the routine "B" next, the statement
C
C   IF (ICFELL('B - UNCLEARED PRIOR ERROR',1).NE.0) RETURN
C
C ensures that the error message from routine "A" will be printed, that
C it will be replaced by "B - UNCLEARED PRIOR ERROR", and that "B" will
C not attempt to execute.
C
C
C The common blocks SECOMI and SECOMC are used to hold shared variables
C of types INTEGER and CHARACTER, respectively, for the routine SETER
C and associated routines.  For descriptions of these variables and for
C default values of them, see the block data routine SEBLDAX.
C
        COMMON /SECOMI/ IERRU,IERRF,IRECF,LOMSG
        SAVE   /SECOMI/
C
        COMMON /SECOMC/ ERMSG
          CHARACTER*256 ERMSG
        SAVE   /SECOMC/
C
C Define a character temporary to use.
C
        CHARACTER*256 CTEMP
C
C Set the value of the function to the current value of the error flag.
C
        ICFELL=IERRF
C
C If the current error flag is non-zero, update its value and the value
C of the error message as directed by the values of the arguments of
C ICFELL.
C
        IF (IERRF.NE.0) THEN
C
          IF (MESSG.NE.' ') THEN
C
            LENMSG=ICLOEM(MESSG)
C
            IF (LENMSG.LE.6) THEN
              CTEMP=ERMSG
              ERMSG=MESSG(1:LENMSG)//'/'//CTEMP
            ELSE
              CALL EPRIN
              ERMSG=MESSG
            END IF
C
            LOMSG=ICLOEM(ERMSG)
C
          END IF
C
          IF (NERRF.NE.0) IERRF=NERRF
C
        END IF
C
C Done.
C
        RETURN
C
      END

File: ncarg2d/man/error_handling/icfell.m
.TH ICFELL 3NCARG "March 1994" UNIX "NCAR GRAPHICS"
.na
.nh
.SH NAME
ICFELL - Called to check for existing error conditions and (perhaps) to update
the current error message and error flag.
.SH SYNOPSIS
NERR = ICFELL(MESSG,NERRF)
.SH C-BINDING SYNOPSIS
#include <ncarg/ncargC.h>
.sp
int c_icfell(char *messg, int nerrf)
.SH DESCRIPTION 
ICFELL (which stands for "I Check For Errors on Lower Level") is used to
check for the occurrence of a recoverable error in a lower-level routine and
(perhaps) to update the current internal error message and error flag.
.sp
The old value of the error flag is returned as the value of the function
ICFELL. If that value is zero, nothing else has been done. If the value of
ICFELL is non-zero, the following actions have been taken:
.IP " -" 4
If MESSG is blank, the current error message has not been changed.
.IP " -" 4
If MESSG is non-blank and its length is 6 or less, it should be the name of
the routine referencing ICFELL; the current error message has been altered by
prepending first a slash and then MESSG. (Note that, if an error occurs several
levels deep, the effect of using ICFELL in this manner at each level is,
effectively, to generate traceback information in the error message.)
.IP " -" 4
If MESSG is non-blank and its length is 7 or greater, its value has become
the new value of the current error message and the previous error message
has been printed. This is intended for use at the beginning of an NCAR
Graphics routine to check for an outstanding error that the user has not
recovered from and to ensure that the message for the outstanding error
gets printed.
.IP " -" 4
If the expression NERRF has the value zero, the current error flag has not
been changed.
.IP " -" 4
If the expression NERRF has a non-zero value, the value of the current error
flag has been made equal to that value.
.PP
An example: Assume that the routine "A" calls the routine "B" and that "B"
detects an error and calls SETER with error number "32" and error message
"B - ERROR HAS OCCURRED". If recovery mode is not in effect, SETER prints
the error message and STOPs; if recovery mode is in effect, control returns
from "SETER" to "B" and thence to "A". At that point, the statement
"IF (ICFELL('A',13).NE.0) RETURN" detects the fact that an error has occurred
in "B" and results in a return from "A" to whatever routine called it. It also
changes the current error message to read 'A/B - ERROR HAS OCCURRED' and
changes the error number from "32" to "13".
.PP
Another example: Assume that the NCAR Graphics routine "A" is called when
recovery mode is set and that it detects an error, calls SETER, and RETURNs
to the user. If the user neglects to check the error state and calls the
routine "B" next, the statement "IF (ICFELL('B - UNCLEARED PRIOR ERROR',1).NE.0) RETURN"
ensures that the error message from routine "A" will be printed, that it will
be replaced by an error message referring to the routine "B", and that "B"
won't do anything else.
.sp
The arguments of ICFELL are as follows:
.sp
.IP "MESSG" 12
(an input variable or constant of type CHARACTER) - A character string saying
what is to be done to the current error message (when an error has occurred).
If MESSG is blank, no change is to be made in the current error message. If
the length of MESSG is six characters or fewer, a slash is prepended to the
current error message and then MESSG is prepended to that. If the length of
MESSG is seven characters or more, MESSG is a complete new error message to
be substituted for the current one; it should be of the form
"XXXXXX - TEXT DESCRIBING THE ERROR", where "XXXXXX" is the name of the
routine in which the error occurred, and it should not be more than 113
characters long.
.IP "NERRF" 12
(an input expression of type INTEGER) - A value in the range from -999 to
+9999. If this value is zero, no change is made in the internal error flag;
otherwise, it is the desired new value of that flag (when an error has
occurred).
.SH C-BINDING DESCRIPTION 
The C-binding argument descriptions are the same as the FORTRAN 
argument descriptions.
.SH EXAMPLES
Use the ncargex command to see the following relevant
examples: 
tseter,
arex02.
.SH ACCESS
To use ICFELL or c_icfell, load the NCAR Graphics libraries ncarg, ncarg_gks,
and ncarg_c, preferably in that order.
.SH SEE ALSO
Online:
entsr, eprin, errof, error_handling, fdum, icloem, nerro, retsr,
semess, seter, ncarg_cbind
.SH COPYRIGHT
Copyright (C) 1987-2009
.br
University Corporation for Atmospheric Research
.br
The use of this Software is governed by a License Agreement.
